File: backend/CattleTrace/api/v1/serializers/health.py
"""Health record serializers."""

from rest_framework import serializers

from CattleTrace.models import Disease, HealthRecord, Vaccine


class DiseaseSerializer(serializers.ModelSerializer):
    class Meta:
        model = Disease
        fields = ('id', 'name', 'description', 'is_notifiable')


class VaccineSerializer(serializers.ModelSerializer):
    class Meta:
        model = Vaccine
        fields = ('id', 'name', 'manufacturer', 'validity_days')


class HealthRecordSerializer(serializers.ModelSerializer):
    diagnosis_detail = DiseaseSerializer(source='diagnosis', read_only=True)
    vaccine_used_detail = VaccineSerializer(source='vaccine_used', read_only=True)
    vet = serializers.PrimaryKeyRelatedField(read_only=True)
    vet_name = serializers.SerializerMethodField()
    animal_tag = serializers.CharField(source='animal.tag_number', read_only=True)
    credibility_level_display = serializers.CharField(source='get_credibility_level_display', read_only=True)

    class Meta:
        model = HealthRecord
        fields = (
            'id',
            'animal',
            'animal_tag',
            'record_type',
            'date',
            'vet',
            'vet_name',
            'credibility_level',
            'credibility_level_display',
            'is_escalated',
            'diagnosis',
            'diagnosis_detail',
            'vaccine_used',
            'vaccine_used_detail',
            'medication',
            'dosage',
            'next_due_date',
            'temperature',
            'notes',
            'certificate_no',
            'created_at',
        )
        read_only_fields = ('id', 'vet', 'vet_name', 'credibility_level', 'is_escalated', 'created_at')

    def get_vet_name(self, obj):
        if not obj.vet:
            return None
        full_name = obj.vet.get_full_name().strip()
        return full_name or obj.vet.username

    def create(self, validated_data):
        user = self.context['request'].user
        if user.role in (user.Role.VET, user.Role.CAHW):
            validated_data['vet'] = user
        if user.role == user.Role.VET:
            validated_data['credibility_level'] = HealthRecord.CredibilityLevel.VET_VERIFIED
        elif user.role == user.Role.CAHW:
            validated_data['credibility_level'] = HealthRecord.CredibilityLevel.CAHW_OBSERVATION
        else:
            validated_data['credibility_level'] = HealthRecord.CredibilityLevel.SELF_REPORTED
        return super().create(validated_data)

File: frontend/src/app/lib/api/health.ts
import type { HealthEvent } from "../types";
import type { PaginatedResponse } from "./animals";

export type ApiHealthRecordType =
  | "vaccination"
  | "treatment"
  | "examination"
  | "deworming"
  | "dipping"
  | "other";

export interface ApiDisease {
  id: number;
  name: string;
  description: string;
  is_notifiable: boolean;
}

export interface ApiVaccine {
  id: number;
  name: string;
  manufacturer: string;
  validity_days: number;
}

export interface ApiHealthRecord {
  id: number;
  animal: number;
  animal_tag: string;
  record_type: ApiHealthRecordType;
  date: string;
  vet: number | null;
  diagnosis: number | null;
  diagnosis_detail: ApiDisease | null;
  vaccine_used: number | null;
  vaccine_used_detail: ApiVaccine | null;
  medication: string;
  dosage: string;
  next_due_date: string | null;
  temperature: string | null;
  notes: string;
  certificate_no: string;
  created_at: string;
}

export interface HealthRecordListParams {
  page?: number;
  pageSize?: number;
  search?: string;
  animal?: string;
  disease?: string;
  ordering?: "date" | "-date" | "created_at" | "-created_at";
}

export interface HealthRecordPayload {
  animal: number;
  record_type: ApiHealthRecordType;
  date: string;
  diagnosis?: number | null;
  vaccine_used?: number | null;
  medication?: string;
  dosage?: string;
  next_due_date?: string | null;
  temperature?: string | null;
  notes?: string;
  certificate_no?: string;
}

export type HealthRecordListResponse = PaginatedResponse<ApiHealthRecord> | ApiHealthRecord[];

function getEventType(record: ApiHealthRecord): HealthEvent["eventType"] {
  if (record.record_type === "vaccination") return "Vaccination";
  if (record.diagnosis_detail) return "Disease";
  if (record.record_type === "other") return "Treatment";
  return "Treatment";
}

export function mapApiHealthRecordToHealthEvent(record: ApiHealthRecord): HealthEvent {
  const eventType = getEventType(record);
  const disease = record.diagnosis_detail?.name || (eventType === "Disease" ? record.medication : undefined);
  const vaccine = record.vaccine_used_detail?.name || (eventType === "Vaccination" ? record.medication : undefined);

  return {
    id: String(record.id),
    animalId: String(record.animal),
    animalRfid: record.animal_tag,
    eventType,
    disease,
    vaccine,
    date: record.date,
    recordedBy: record.vet ? `Vet #${record.vet}` : "CattleTrace",
    credentialLevel: record.vet ? "Veterinary record" : "Authorized record",
    notes: record.notes || undefined,
    severity: eventType === "Disease" ? "Medium" : "Low",
  };
}
